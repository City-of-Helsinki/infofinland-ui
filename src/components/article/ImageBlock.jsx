import Block from '@/components/layout/Block'
import Image from 'next/image'
import { useTranslation } from 'next-i18next'

const getFit = ({ width, height }) => (width < height ? 'contain' : 'cover')

export const ArticleImage = ({
  src,
  height,
  width,
  title,
  alt,
  caption,
  photographer,
  className,
  fit,
  layout = 'fill',
}) => {
  const { t } = useTranslation('common')
  return (
    <>
      <div className={className}>
        {src && (
          <Image
            alt={alt}
            src={src}
            title={title}
            layout={layout}
            objectFit={fit || getFit({ width, height })}
            {...(layout === 'fill' ? {} : { width, height })}
          />
        )}
      </div>

      {caption && (
        <p className="mt-2 text-body-small text-gray-dark">{caption}</p>
      )}
      {photographer && (
        <p className="mt-2 text-action text-gray-dark">
          {t('images.photographer')}: {photographer}
        </p>
      )}
    </>
  )
}

const ImageBlock = ({ caption, src, height, width, title, alt }) => {
  return (
    <Block className="my-16" hero>
      <ArticleImage
        className="overflow-hidden relative bg-gray-lighter-teal rounded aspect-video"
        alt={alt}
        src={src}
        width={width}
        height={height}
        caption={caption}
        title={title}
      />
    </Block>
  )
}

export default ImageBlock
