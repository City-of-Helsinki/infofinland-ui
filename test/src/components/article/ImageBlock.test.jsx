/**
 * @jest-environment jsdom
 */

import React from 'react'
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react'
import ImageBlock, { ArticleImage } from '@/components/article/ImageBlock'

describe('ImageBlock', () => {
  it('should render the given image inside a Block', () => {
    const { asFragment } = render(
      <ImageBlock
        alt="alt"
        src="/foo.png"
        height={100}
        width={200}
        caption="caption"
        title="title"
      />
    )
    // we can only use toBeInTheDocument because it was imported
    // in the jest.setup.js and configured in jest.config.js
    expect(asFragment()).toMatchSnapshot()
  })
})

describe('ArticleImage', () => {
  it('should render image alt, title, className and and caption', () => {
    const { asFragment } = render(
      <ArticleImage
        alt="alt"
        src="/foo.png"
        caption="caption-big"
        title="title-big"
        className="test-class"
      />
    )
    // we can only use toBeInTheDocument because it was imported
    // in the jest.setup.js and configured in jest.config.js
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render use layout="fill" as default and ignore heigh and width', () => {
    const { asFragment } = render(
      <ArticleImage src="/foo.png" height={103} width={404} />
    )
    // we can only use toBeInTheDocument because it was imported
    // in the jest.setup.js and configured in jest.config.js
    expect(asFragment()).toMatchSnapshot()
  })

  it('should use objectFit="contain" if width < height', () => {
    const { asFragment } = render(
      <ArticleImage src="/foo.png" height={503} width={404} />
    )
    // we can only use toBeInTheDocument because it was imported
    // in the jest.setup.js and configured in jest.config.js
    expect(asFragment()).toMatchSnapshot()
  })

  it('should use objectFit="hover" if width >= height', () => {
    const { asFragment } = render(
      <ArticleImage src="/foo.png" height={203} width={404} />
    )
    // we can only use toBeInTheDocument because it was imported
    // in the jest.setup.js and configured in jest.config.js
    expect(asFragment()).toMatchSnapshot()
  })

  it('should require height and width for other layouts and throw an error if they are missing', () => {
    //responsive, intrinsic, fixed, fill,
    const { asFragment } = render(
      <ArticleImage
        src="/foo.png"
        height={103}
        width={404}
        layout="responsive"
      />
    )

    // we can only use toBeInTheDocument because it was imported
    // in the jest.setup.js and configured in jest.config.js
    expect(asFragment()).toMatchSnapshot()

    expect(() =>
      render(<ArticleImage src="/foo.png" layout="responsive" />)
    ).toThrow()

    expect(() =>
      render(<ArticleImage src="/foo.png" layout="intrinsic" />)
    ).toThrow()

    expect(() =>
      render(<ArticleImage src="/foo.png" layout="fixed" />)
    ).toThrow()
  })
})
