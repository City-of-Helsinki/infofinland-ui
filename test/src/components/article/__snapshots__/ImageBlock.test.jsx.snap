// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleImage should render image alt, title, className and and caption 1`] = `
<DocumentFragment>
  <div
    class="test-class ifu-loader__rings"
  >
    <div
      class="flex absolute justify-items-center items-center w-full h-full"
    >
      <div
        class="scale-75 lg:scale-100 ifu-loader__image--ripple"
      >
        <div />
        <div />
      </div>
    </div>
    <span
      style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
    >
      <img
        alt="alt"
        data-nimg="fill"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
        title="title-big"
      />
      <noscript />
    </span>
  </div>
  <p
    class="mt-2 text-body-small text-gray-dark"
  >
    caption-big
  </p>
</DocumentFragment>
`;

exports[`ArticleImage should render use layout="fill" as default and ignore heigh and width 1`] = `
<DocumentFragment>
  <div
    class="ifu-loader__rings"
  >
    <div
      class="flex absolute justify-items-center items-center w-full h-full"
    >
      <div
        class="scale-75 lg:scale-100 ifu-loader__image--ripple"
      >
        <div />
        <div />
      </div>
    </div>
    <span
      style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
    >
      <img
        data-nimg="fill"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
      />
      <noscript />
    </span>
  </div>
</DocumentFragment>
`;

exports[`ArticleImage should require height and width for other layouts and throw an error if they are missing 1`] = `
<DocumentFragment>
  <div
    class="ifu-loader__rings"
  >
    <div
      class="flex absolute justify-items-center items-center w-full h-full"
    >
      <div
        class="scale-75 lg:scale-100 ifu-loader__image--ripple"
      >
        <div />
        <div />
      </div>
    </div>
    <span
      style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
    >
      <span
        style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 25.495049504950494% 0px 0px 0px;"
      />
      <img
        data-nimg="responsive"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
      />
      <noscript />
    </span>
  </div>
</DocumentFragment>
`;

exports[`ArticleImage should use objectFit="contain" if width < height 1`] = `
<DocumentFragment>
  <div
    class="ifu-loader__rings"
  >
    <div
      class="flex absolute justify-items-center items-center w-full h-full"
    >
      <div
        class="scale-75 lg:scale-100 ifu-loader__image--ripple"
      >
        <div />
        <div />
      </div>
    </div>
    <span
      style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
    >
      <img
        data-nimg="fill"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
      />
      <noscript />
    </span>
  </div>
</DocumentFragment>
`;

exports[`ArticleImage should use objectFit="hover" if width >= height 1`] = `
<DocumentFragment>
  <div
    class="ifu-loader__rings"
  >
    <div
      class="flex absolute justify-items-center items-center w-full h-full"
    >
      <div
        class="scale-75 lg:scale-100 ifu-loader__image--ripple"
      >
        <div />
        <div />
      </div>
    </div>
    <span
      style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
    >
      <img
        data-nimg="fill"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
      />
      <noscript />
    </span>
  </div>
</DocumentFragment>
`;

exports[`ImageBlock should render the given image inside a Block 1`] = `
<DocumentFragment>
  <div
    class="mx-6 lg:mx-12 xl:mx-28 2xl:mx-48 3xl:ms-64 3xl:max-w-4xl my-16"
  >
    <div
      class="overflow-hidden relative bg-gray-white rounded aspect-video ifu-loader__rings"
    >
      <div
        class="flex absolute justify-items-center items-center w-full h-full"
      >
        <div
          class="scale-75 lg:scale-100 ifu-loader__image--ripple"
        >
          <div />
          <div />
        </div>
      </div>
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
      >
        <img
          alt="alt"
          data-nimg="fill"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
          title="title"
        />
        <noscript />
      </span>
    </div>
    <p
      class="mt-2 text-body-small text-gray-dark"
    >
      caption
    </p>
  </div>
</DocumentFragment>
`;
